cmd/kubeadm/app/util/runtime/runtime.go

```diff
@@ -180,3 +181,52 @@ func (runtime *DockerRuntime) ImageExists(image string) (bool, error) {
 	err := runtime.exec.Command("docker", "inspect", image).Run()
 	return err == nil, nil
 }
+
+// isExistingSocket checks if path exists and is domain socket
+func isExistingSocket(path string) bool {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+
+	return fileInfo.Mode()&os.ModeSocket != 0
+}
+
+// detectCRISocketImpl is separated out only for test purposes, DON'T call it directly, use DetectCRISocket instead
+func detectCRISocketImpl(isSocket func(string) bool) (string, error) {
+	foundCRISockets := []string{}
+	knownCRISockets := []string{
+		// /var/run/dockershim.sock is deliberately omitted from here, we have a separate Docker check below
+		"/run/containerd/containerd.sock",
+		"/var/run/crio/crio.sock",
+	}
+
+	const dockerSocket = "/var/run/docker.sock" // The Docker socket is not CRI compatible
+	if isSocket(dockerSocket) {
+		foundCRISockets = append(foundCRISockets, kubeadmapiv1beta1.DefaultDockerCRISocket)
+	}
+
+	for _, socket := range knownCRISockets {
+		if isSocket(socket) {
+			foundCRISockets = append(foundCRISockets, socket)
+		}
+	}
+
+	switch len(foundCRISockets) {
+	case 0:
+		return "", pkgerrors.New("No CRI socket detected, please use the --cri-socket option")
+	case 1:
+		return foundCRISockets[0], nil
```

Comment:
Does this mean that if docker is running this function will always return /var/run/docker.sock ? Would it be better to change the order of known sockets? If CRI-O or containerd is configured and running should we prefer those even if docker is running?

Reply:
No, this means, that if you have only one runtime environment (no matter if Docker or CRI socket backed one) it will use that. If there is no runtime environment detected or multiple ones are detected (for example Docker & CRI-O) it will display error message and force the user to use --cri-socket to supply the socket.